bundles/mapping/mapmodule/plugin/getinfo/GetInfoPlugin.js

```diff
@@ -385,6 +387,111 @@ Oskari.clazz.define(
             me._showGfiInfo = callback;
         },
 
+        /**
+         * Sends GFIResultEvent.
+         *
+         * @method _sendGFIResultEvent
+         * @private
+         * @param  {Object} data
+         */
+        _sendGFIResultEvent: function (data) {
+            /**
+             * Checks at if data has JSON
+             * @param {Object} data
+             * @returns {Boolean} true if data has JSON Object, other false
+             */
+            const hasJSON = (data) => {
+                try {
+                    if (typeof data === 'string') {
+                        JSON.parse(data);
+                        return true;
+                    } else if (typeof data === 'object') {
+                        return true;
+                    }
+                    return false;
+                } catch (e) {}
+                return false;
+            };
+            /**
+             * Checks at if data has GeoJSON
+             * @param {Object} data
+             * @returns {Boolean} true if data has GeoJSON Object, other false
+             */
+            const hasGeoJSON = (data) => {
+                try {
+                    new GeoJSON().readFeatures(data);
+                } catch (e) {
+                    return false;
+                }
+                return true;
+            };
+
+            /**
+             * Gets GFI response type
+             * @param {Object} feature
+             * @returns {String} GFI response type (json, geojson or text)
+             */
+            const getGfiResponseType = (feature) => {
+                if (hasJSON(feature.content) && !hasGeoJSON(feature.content)) {
+                    return 'json';
+                } else if (hasJSON(feature.content) && hasGeoJSON(feature.content)) {
+                    return 'geojson';
+                } else if (typeof feature === 'object' && !hasGeoJSON(feature)) {
+                    return 'json';
+                } else if (typeof feature === 'object' && hasGeoJSON(feature)) {
+                    return 'geojson';
+                }
+                return 'text';
```

Comment:
Reduce calls to parsing functions:
```suggestion
            const getGfiResponseType = (feature) => {
                const isJSON = hasJSON(feature.content);
                if (isJSON) {
                    if (hasGeoJSON(feature.content)) {
                        return 'geojson';
                    }
                    return 'json';
                }
                return 'text';
```

Reply:
How about WFS feature clicks ? Now suggested change not sends this event out when clicking WFS feature, there clicked feature attributes data come directly `feature` object. 

For example: 
```javascript
feature = { 
    'Attribute1': 'aaa',
    'Attribute2' 'bbb'
};
```

Or keep this event only the GFI-stuff that is fetched through the server ?

Reply:
True, are there any other exceptions of could we do something like:
```suggestion
            const isWMSgfi = (feature) => feature.type = 'wmslayer' &&
                ['presentationType', 'content', 'type'].every(key => !!feature[key]);
            const getGfiResponseType = (feature) => {                
                if (isWMSgfi(feature)) {
                    const isJSON = hasJSON(feature.content);
                    if (isJSON) {
                        if (hasGeoJSON(feature.content)) {
                            return 'geojson';
                        }
                        return 'json';
                    }
                    return 'text';
                } else {
                    // Content is WFS feature properties
                    return 'json';
                }
            }
```

Reply:
BTW these helper-functions should be moved out from this function so we don't need to declare them for each call. Since they can operate on just the params the receive as input they can be moved.

Reply:
Try reducing the JSON.parse() calls in general. Since now it's done to detect if the content is JSON. Then again to parse the content to an object etc.